java/arcs/core/analysis/DependencyNode.kt

```diff
@@ -84,56 +84,150 @@ private typealias Path = List<Identifier>
  */
 sealed class DependencyNode {
 
+  /** Path of [Identifier]s representing access to a handle connection. */
+  abstract val path: Path
+
+  /** Association with the current node and child nodes, via the children's Identifier. */
+  abstract val associations: Map<Identifier, DependencyNode>
+
+  /** Set of [DependencyNode]s that the current node depends on. */
+  abstract val dependency: Set<DependencyNode>
+
+  /** Set of [DependencyNode]s that bear influence on the current node. */
+  abstract val influencedBy: Set<DependencyNode>
+
+  /** Gets the [Identifier] of the current node, if present (`null` otherwise). */
+  val id: Identifier?
+    get() = path.last()
+
+  /** Expresses influence on to the current [DependencyNode]. */
+  abstract fun influencedBy(influence: Set<DependencyNode>): DependencyNode
+
+  /** Replace the associations with new mappings. */
+  abstract fun add(vararg other: Pair<Identifier, DependencyNode>): DependencyNode
+
+  /** Returns the [DependencyNode] associated with the input [Identifier]. */
+  fun lookup(key: Identifier): DependencyNode = requireNotNull(associations[key]) {
+    "Identifier '$key' is not found in `associations`."
+  }
+
   /** An unmodified input (from a handle connection) used in a Paxel [Expression]. */
-  data class Input(val path: Path = emptyList()) : DependencyNode() {
-    constructor(vararg fields: Identifier) : this(listOf(*fields))
+  data class Input private constructor(
```

Comment:
A difficulty in implementing the high level feedback (above): `data class`es cannot be subclassed. Thus, `Input` is a sibling class for `AssociationNode`, not a child. To make these two more compatible, I pulled `associations` up as an abstract field on `DependencyNode`. 
I'm curious what you think about this approach (the abstract goal, or this particular implementation).

Reply:
I was actually suggesting that `AssociationNode` and `Input` can be one and the same, where `associations` will be empty for `Input`